// @vtaits/react-color-picker/lib/HueSpectrum.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _common = _interopRequireWildcard(require("./utils/common"));

var _defaultColor = _interopRequireDefault(require("./defaultColor"));

var _validate = _interopRequireDefault(require("./utils/validate"));

var _toColorValue = _interopRequireDefault(require("./utils/toColorValue"));

var HueSpectrum =
/*#__PURE__*/
function (_BaseComponent) {
  (0, _inherits2["default"])(HueSpectrum, _BaseComponent);

  function HueSpectrum() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2["default"])(this, HueSpectrum);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2["default"])(this, (_getPrototypeOf2 = (0, _getPrototypeOf3["default"])(HueSpectrum)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "state", (0, _objectSpread2["default"])({}, _common.baseInitialState, {
      h: 0
    }));
    return _this;
  }

  (0, _createClass2["default"])(HueSpectrum, [{
    key: "isComponentMounted",
    value: function isComponentMounted() {
      return this.mounted;
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      this.mounted = true;
      this.updateDragPositionIf();
    }
  }, {
    key: "updateDragPositionIf",
    value: function updateDragPositionIf() {
      if (!this.props.height) {
        this.setState({});
      }
    }
  }, {
    key: "getDragPosition",
    value: function getDragPosition() {
      var _this$props = this.props,
          height = _this$props.height,
          pointerSize = _this$props.pointerSize;

      if (!height && !this.isComponentMounted()) {
        return null;
      }

      var computedHeight = height || this.getDOMRegion().getHeight();
      var size = pointerSize;
      var pos = Math.round(this.hsv.h * computedHeight / 360);
      var diff = Math.round(size / 2);
      return pos - diff;
    }
  }, {
    key: "updateColor",
    value: function updateColor(point) {
      var newPoint = (0, _validate["default"])(point);
      this.hsv.h = newPoint.y * 360 / newPoint.height;
      var newHsv = (0, _objectSpread2["default"])({}, this.hsv);
      var newH;

      if (this.hsv.h !== 0) {
        newH = this.hsv.h;
      }

      newH = this.hsv.h !== 0 ? this.hsv.h : 0;
      this.setState({
        h: newH
      });
      this.hsv = newHsv;
      return newHsv;
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          style = _this$props2.style,
          value = _this$props2.value,
          defaultValue = _this$props2.defaultValue,
          defaultColor = _this$props2.defaultColor,
          pointerSize = _this$props2.pointerSize,
          height = _this$props2.height,
          width = _this$props2.width;
      var _this$state = this.state,
          stateValue = _this$state.value,
          h = _this$state.h;
      this.hsv = (0, _toColorValue["default"])(stateValue || value || defaultValue || defaultColor);

      if (h === 360 && !this.hsv.h) {
        // in order to show bottom red as well on drag
        this.hsv.h = 360;
      }

      var rootStyle = (0, _objectSpread2["default"])({}, style);

      if (height) {
        rootStyle.height = height;
      }

      if (width) {
        rootStyle.width = width;
      }

      var dragStyle = {
        height: pointerSize
      };
      var dragPos = this.getDragPosition();

      if (dragPos !== null) {
        dragStyle.top = dragPos;
        dragStyle.display = 'block';
      }

      return _react["default"].createElement("div", {
        className: "react-color-picker__hue-spectrum",
        style: rootStyle,
        onMouseDown: this.onMouseDown,
        role: "button",
        ref: this.rootRef,
        tabIndex: 0
      }, _react["default"].createElement("div", {
        className: "react-color-picker__hue-drag",
        style: dragStyle
      }, _react["default"].createElement("div", {
        className: "react-color-picker__hue-inner"
      })));
    }
  }]);
  return HueSpectrum;
}(_common["default"]);

(0, _defineProperty2["default"])(HueSpectrum, "propTypes", (0, _objectSpread2["default"])({}, _common.basePropTypes, {
  height: _propTypes["default"].number,
  width: _propTypes["default"].number,
  pointerSize: _propTypes["default"].number,
  defaultColor: _propTypes["default"].any,
  isHueSpectrum: _propTypes["default"].bool
}));
(0, _defineProperty2["default"])(HueSpectrum, "defaultProps", (0, _objectSpread2["default"])({}, _common.baseDefaultProps, {
  height: 300,
  width: 30,
  pointerSize: 3,
  defaultColor: _defaultColor["default"],
  isHueSpectrum: true
}));
var _default = HueSpectrum;
exports["default"] = _default;
