// @vtaits/react-color-picker/lib/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "HueSpectrum", {
  enumerable: true,
  get: function get() {
    return _HueSpectrum["default"];
  }
});
Object.defineProperty(exports, "SaturationSpectrum", {
  enumerable: true,
  get: function get() {
    return _SaturationSpectrum["default"];
  }
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _HueSpectrum = _interopRequireDefault(require("./HueSpectrum"));

var _SaturationSpectrum = _interopRequireDefault(require("./SaturationSpectrum"));

var _toColorValue = _interopRequireDefault(require("./utils/toColorValue"));

var _toStringValue = _interopRequireDefault(require("./utils/toStringValue"));

var _defaultColor = _interopRequireDefault(require("./defaultColor"));

var ColorPicker =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2["default"])(ColorPicker, _Component);

  function ColorPicker(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, ColorPicker);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(ColorPicker).call(this, props));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleSaturationChange", function (color) {
      _this.handleChange(color);
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleHueChange", function (color) {
      _this.handleChange(color);
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleHueDrag", function (hsv) {
      _this.setState({
        dragHue: hsv.h
      });

      _this.handleDrag(hsv);
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleSaturationDrag", function (hsv) {
      _this.handleDrag(hsv);
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleHueMouseDown", function (hsv) {
      _this.setState({
        dragHue: hsv.h
      });
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleSaturationMouseDown", function (hsv) {
      _this.setState({
        dragHue: hsv.h
      });
    });
    var defaultValue = props.defaultValue;
    _this.state = {
      value: defaultValue
    };
    return _this;
  }

  (0, _createClass2["default"])(ColorPicker, [{
    key: "handleDrag",
    value: function handleDrag(color) {
      var _this$props = this.props,
          value = _this$props.value,
          onDrag = _this$props.onDrag;

      if (!value) {
        this.setState({
          value: color
        });
      }

      onDrag((0, _toStringValue["default"])(color), color);
    }
  }, {
    key: "handleChange",
    value: function handleChange(color) {
      var onChange = this.props.onChange;
      this.setState({
        dragHue: null
      });
      var newColor = (0, _objectSpread2["default"])({}, color);
      var value = (0, _toStringValue["default"])(newColor);
      onChange(value, newColor);
    }
  }, {
    key: "render",
    value: function render() {
      var props = this.props;
      var color = props.color,
          propsClassName = props.className,
          propsHueStyle = props.hueStyle,
          hueHeight = props.hueHeight,
          hueMargin = props.hueMargin,
          hueWidth = props.hueWidth,
          propsDefaultValue = props.defaultValue,
          defaultColor = props.defaultColor,
          propsValue = props.value,
          saturationHeight = props.saturationHeight,
          saturationWidth = props.saturationWidth,
          divProps = (0, _objectWithoutProperties2["default"])(props, ["color", "className", "hueStyle", "hueHeight", "hueMargin", "hueWidth", "defaultValue", "defaultColor", "value", "saturationHeight", "saturationWidth"]);
      var _this$state = this.state,
          stateValue = _this$state.value,
          dragHue = _this$state.dragHue;
      var className = (0, _classnames["default"])(propsClassName, 'cp react-color-picker');
      var hueStyle = (0, _objectSpread2["default"])({}, propsHueStyle, {
        marginLeft: hueMargin
      });
      var value = props.value ? (0, _toColorValue["default"])(propsValue) : null;
      var defaultValue = !value ? (0, _toColorValue["default"])(stateValue || propsDefaultValue || defaultColor) : null;
      var children = props.children;
      var hueSpectrumProps = null;
      var saturationSpectrumProps = null;

      if (children) {
        children = _react["default"].Children.toArray(children).forEach(function (child) {
          if (child && child.props) {
            if (child.props.isHueSpectrum) {
              hueSpectrumProps = (0, _objectSpread2["default"])({}, child.props);
            }

            if (child.props.isSaturationSpectrum) {
              saturationSpectrumProps = (0, _objectSpread2["default"])({}, child.props);
            }
          }
        });
      }

      var saturationConfig = (0, _objectSpread2["default"])({
        onDrag: this.handleSaturationDrag,
        onChange: this.handleSaturationChange,
        onMouseDown: this.handleSaturationMouseDown
      }, saturationSpectrumProps);

      if (saturationConfig.width === undefined) {
        saturationConfig.width = saturationWidth;
      }

      if (saturationConfig.height === undefined) {
        saturationConfig.height = saturationHeight;
      }

      saturationConfig.inPicker = true;
      var hueConfig = (0, _objectSpread2["default"])({
        onDrag: this.handleHueDrag,
        onChange: this.handleHueChange,
        onMouseDown: this.handleHueMouseDown,
        style: hueStyle
      }, hueSpectrumProps);

      if (hueConfig.width === undefined) {
        hueConfig.width = hueWidth;
      }

      if (hueConfig.height === undefined) {
        hueConfig.height = hueHeight || saturationHeight;
      }

      hueConfig.inPicker = true;

      if (dragHue) {
        (value || defaultValue).h = dragHue;
      } // both value and defaultValue are objects like: {h,s,v}


      if (value) {
        saturationConfig.value = (0, _objectSpread2["default"])({}, value);
        hueConfig.value = (0, _objectSpread2["default"])({}, value);
      } else {
        saturationConfig.defaultValue = (0, _objectSpread2["default"])({}, defaultValue);
        hueConfig.defaultValue = (0, _objectSpread2["default"])({}, defaultValue);
      }

      return _react["default"].createElement("div", (0, _extends2["default"])({}, divProps, {
        className: className
      }), _react["default"].createElement(_SaturationSpectrum["default"], saturationConfig), _react["default"].createElement(_HueSpectrum["default"], hueConfig));
    }
  }]);
  return ColorPicker;
}(_react.Component);

(0, _defineProperty2["default"])(ColorPicker, "propTypes", {
  onDrag: _propTypes["default"].func,
  onChange: _propTypes["default"].func,
  value: _propTypes["default"].any,
  defaultValue: _propTypes["default"].any,
  defaultColor: _propTypes["default"].any,
  hueHeight: _propTypes["default"].number,
  hueMargin: _propTypes["default"].number,
  hueWidth: _propTypes["default"].number,
  saturationWidth: _propTypes["default"].number,
  saturationHeight: _propTypes["default"].number
});
(0, _defineProperty2["default"])(ColorPicker, "defaultProps", {
  onDrag: Function.prototype,
  onChange: Function.prototype,
  value: null,
  defaultValue: null,
  defaultColor: _defaultColor["default"],
  hueHeight: null,
  hueMargin: 10,
  hueWidth: 30,
  saturationWidth: 300,
  saturationHeight: 300
});
var _default = ColorPicker;
exports["default"] = _default;
