// @vtaits/react-color-picker/lib/SaturationSpectrum.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _color = require("./utils/color");

var _common = _interopRequireWildcard(require("./utils/common"));

var _defaultColor = _interopRequireDefault(require("./defaultColor"));

var _validate = _interopRequireDefault(require("./utils/validate"));

var _toColorValue = _interopRequireDefault(require("./utils/toColorValue"));

var getSaturationForPoint = function getSaturationForPoint(point) {
  return point.x / point.width;
};

var getColorValueForPoint = function getColorValueForPoint(point) {
  return (point.height - point.y) / point.height;
};

var prepareBackgroundColor = function prepareBackgroundColor(color) {
  var hsv = color;
  var col = (0, _color.fromRatio)({
    h: hsv.h % 360 / 360,
    s: 1,
    v: 1
  });
  return col.toRgbString();
};

var SaturationSpectrum =
/*#__PURE__*/
function (_BaseComponent) {
  (0, _inherits2["default"])(SaturationSpectrum, _BaseComponent);

  function SaturationSpectrum() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2["default"])(this, SaturationSpectrum);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2["default"])(this, (_getPrototypeOf2 = (0, _getPrototypeOf3["default"])(SaturationSpectrum)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "state", (0, _objectSpread2["default"])({}, _common.baseInitialState, {
      pointerTop: null,
      pointerLeft: null
    }));
    return _this;
  }

  (0, _createClass2["default"])(SaturationSpectrum, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.mounted = true;
      this.updateDragPositionIf();
    }
  }, {
    key: "isComponentMounted",
    value: function isComponentMounted() {
      return this.mounted;
    }
  }, {
    key: "updateDragPositionIf",
    value: function updateDragPositionIf() {
      if (!this.props.height || !this.props.width) {
        this.setState({});
      }
    }
  }, {
    key: "getDragPosition",
    value: function getDragPosition() {
      var _this$props = this.props,
          value = _this$props.value,
          pointerSize = _this$props.pointerSize;
      var _this$props2 = this.props,
          width = _this$props2.width,
          height = _this$props2.height;
      var mouseDown = this.state.mouseDown;
      var sizeDefined = width && height;

      if (!sizeDefined && !this.isComponentMounted()) {
        return null;
      }

      var region;

      if (!sizeDefined) {
        region = this.getDOMRegion();
        height = height || region.getHeight();
        width = width || region.getWidth();
      }

      var x = this.hsv.s * width;
      var y = height - this.hsv.v * height;
      var size = pointerSize;
      var diff = Math.floor(size / 2);

      if (value && mouseDown && !Number.isNaN(mouseDown.x)) {
        x = mouseDown.x;
      }

      return {
        left: x - diff,
        top: y - diff
      };
    }
  }, {
    key: "prepareProps",
    value: function prepareProps(thisProps, state) {
      var props = (0, _objectSpread2["default"])({}, thisProps);
      var color = state.value || props.value || props.defaultValue || props.defaultColor;
      props.color = color;
      this.hsv = (0, _toColorValue["default"])(color);
      props.style = this.prepareStyle(props);
      props.className = (0, _classnames["default"])(thisProps.className, 'react-color-picker__saturation-spectrum');
      return props;
    }
  }, {
    key: "prepareStyle",
    value: function prepareStyle(props) {
      var style = (0, _objectSpread2["default"])({}, props.style);

      if (props.height) {
        style.height = props.height;
      }

      if (props.width) {
        style.width = props.width;
      }

      style.backgroundColor = prepareBackgroundColor(this.hsv);
      return style;
    }
  }, {
    key: "render",
    value: function render() {
      var pointerSize = this.props.pointerSize;
      var props = this.prepareProps(this.props, this.state);
      var dragStyle = {
        width: pointerSize,
        height: pointerSize
      };
      var dragPos = this.getDragPosition();

      if (dragPos) {
        dragStyle.top = dragPos.top;
        dragStyle.left = dragPos.left;
        dragStyle.display = 'block';
      }

      return _react["default"].createElement("div", {
        className: props.className,
        style: props.style,
        onMouseDown: this.onMouseDown,
        role: "button",
        ref: this.rootRef,
        tabIndex: 0
      }, _react["default"].createElement("div", {
        className: "react-color-picker__saturation-white"
      }, _react["default"].createElement("div", {
        className: "react-color-picker__saturation-black"
      })), _react["default"].createElement("div", {
        className: "react-color-picker__saturation-drag",
        style: dragStyle
      }, _react["default"].createElement("div", {
        className: "react-color-picker__saturation-inner"
      })));
    }
  }, {
    key: "updateColor",
    value: function updateColor(point) {
      var newPoint = (0, _validate["default"])(point);
      this.hsv.s = getSaturationForPoint(newPoint);
      this.hsv.v = getColorValueForPoint(newPoint);
      var newHsv = (0, _objectSpread2["default"])({}, this.hsv);
      this.hsv = newHsv;
      return newHsv;
    }
  }]);
  return SaturationSpectrum;
}(_common["default"]);

(0, _defineProperty2["default"])(SaturationSpectrum, "propTypes", (0, _objectSpread2["default"])({}, _common.basePropTypes, {
  height: _propTypes["default"].number,
  width: _propTypes["default"].number,
  pointerSize: _propTypes["default"].number,
  defaultColor: _propTypes["default"].any,
  isSaturationSpectrum: _propTypes["default"].bool
}));
(0, _defineProperty2["default"])(SaturationSpectrum, "defaultProps", (0, _objectSpread2["default"])({}, _common.baseDefaultProps, {
  height: 300,
  width: 300,
  pointerSize: 7,
  defaultColor: _defaultColor["default"],
  isSaturationSpectrum: true
}));
var _default = SaturationSpectrum;
exports["default"] = _default;
